### PR TITLE
[vioscsi] SendSRB minor refactor

### PR DESCRIPTION
**Pull request by benyamin-codez**
The `virtqueue_add_buf()` routine will return 0 on SUCCESS or otherwise a negative number, usually -28 (`ENOSPC`), i.e no space for buffer. An inline comment is added for edification.

Here we refactor `virtqueue_add_buf()` handling to only process the SRB if `virtqueue_add_buf()` returns SUCCESS. Presently, other positive return codes would be processed.

Here we define `VQ_ADD_BUFFER_SUCCESS` in `vioscsi\helper.h`.

To ensure valid data is reported we also move the trace event on failure to above the call to `CompleteRequest()` and wrap the line for improved readability.

### CODE REVIEW

File: vioscsi/helper.c
Comment (JonKohler) [inline vioscsi/helper.c:109]: On virtqueue_add_buf_split(), can you fix up the comments above the function, so that you say returns VQ_ADD_BUFFER_SUCCESS on success?

File: vioscsi/helper.c
Comment (JonKohler) [inline vioscsi/helper.c:112]: both the split and packed variant either return -ENOSPC or zero, which made me think - I don't think ENOSPC is technically the right thing to do here. 

I'd imagine we should actually use ENOMEM or ENOBUFS, as that describes what is happening better than ENOSPC, which is supposed to be used for disk space, no? (ref https://www.gnu.org/software/libc/manual/html_node/Error-Codes.html)

File: vioscsi/helper.c
Comment (JonKohler) [inline vioscsi/helper.c:123]: See my comment above. ENOSPC means no disk space. ENOBUFS means no buffers and ENOMEM is can't alloc mem

File: vioscsi/helper.c
Comment (JonKohler) [inline vioscsi/helper.c:129]: We're concerned there could be at any point any other sort of error, wouldn't we want to put "res" into this log statement?

File: vioscsi/helper.c
Comment (benyamin-codez) [reply vioscsi/helper.c:109]: Nice catch..!
To `VQ_ADD_BUFFER_SUCCESS` methinks...
I'll clean up that init table too...

Comment (benyamin-codez) [reply vioscsi/helper.c:109]: I could but that would be crossing over into [virtio], and then we would need to probably add `VQ_ADD_BUFFER_SUCCESS` into `virtio_ring.h` instead which on reading the head of the file might be a little fraught. I also don't see any maintainer in CODEOWNERS.

File: vioscsi/helper.c
Comment (benyamin-codez) [reply vioscsi/helper.c:112]: Although usually applying to disk devices, ENOSPC refers to _any_ device. The mnemonic ENOSPC refers to ERROR_NO_SPACE. AFAIK, this is returned from the hypervisor without translation rather than something we set. Typically ENOMEM == ENOBUFS in GNU (per your link) and these relate to allocation of virtual memory, whereas virtio sees a _device_ responding with "I have no space for that".

File: vioscsi/helper.c
Comment (benyamin-codez) [reply vioscsi/helper.c:129]: Yes, absolutely... Will do... 8^d

**Comment by benyamin-codez** [comment]
I'll drop another commit here in coming hours but thought it worth mentioning that I'm expecting the following gets clobbered by PR #1150, so won't do anything with that...

https://github.com/virtio-win/kvm-guest-drivers-windows/blob/281bb7b3c5e216fb08449ec05cfe748462ab1ff5/vioscsi/helper.c#L116-L121

**Comment by benyamin-codez** [comment]
Have pushed the additional commit...
Switching this from draft to review...

**Comment by benyamin-codez** [comment]
Check will fail per:

> I'll drop another commit here in coming hours but thought it worth mentioning that I'm expecting the following gets clobbered by PR #1150, so won't do anything with that...
> 
> https://github.com/virtio-win/kvm-guest-drivers-windows/blob/281bb7b3c5e216fb08449ec05cfe748462ab1ff5/vioscsi/helper.c#L116-L121

...for undeclared identifier `res` (as expected).

Let me know if that's ok or you need something else from me.

**Review by JonKohler — COMMENTED** [review]
LGTM at first glance, thanks for the cleanup

**Comment by kostyanf14** [comment]
Hi @benyamin-codez, please fix build errors.

**Comment by benyamin-codez** [comment]
@kostyanf14 

Thanks I will revert the offending changes and put them in a new PR to review after other merges...

**Comment by benyamin-codez** [comment]
@JonKohler 

Jon, I thought I'd better check the modded trace works properly...

I found we do actually set `-ENOSPC` in `virtqueue_add_buf_split()` and `virtqueue_add_buf_packed()` per, e.g. :

https://github.com/virtio-win/kvm-guest-drivers-windows/blob/22d0908a28f382d02d8d3a10a0ee0631af480fdf/VirtIO/VirtIORing.c#L263

However, it returns `-1` per:

https://github.com/virtio-win/kvm-guest-drivers-windows/blob/22d0908a28f382d02d8d3a10a0ee0631af480fdf/VirtIO/osdep.h#L20

I'm still of the view that `ENOSPC` is appropriate here.

@YanVugenfirer

Per the above, do you know if there was any reason we define `ENOSPC` as 1 in `osdep.h` rather than 28...?
I see it was implemented in 2014 per 264f1b6c86f5eeca5e2c9fbd24e3de8dbd0bee1d.

In Windows, 0x1 is `ERROR_INVALID_FUNCTION` and 0x1C (28) is `ERROR_OUT_OF_PAPER`...
Given why and how we are using it, alternatives might include one of the following:
```
ERROR_ACCESS_DENIED     0x5
ERROR_NOT_ENOUGH_MEMORY 0x8
ERROR_NOT_READY         0x15 (21)
ERROR_BUSY              0xAA (170)
```
...but these would require single line changes to VirtIO ( both `VirtIORing.c` and `VirtIORing-Packed.c`).

AFAICT, these are the only files in the tree which reference `ENOSPC`, but deleting the definition from `osdep.h` will result in `undeclared identifier` compilation errors. Perhaps I should include a commit to change it to 28 in this PR (or maybe a separate PR would be best)?

An alternative might be for me to use:

```c
RhelDbgPrint(TRACE_LEVEL_WARNING, 
             " Could not put an SRB into a VQ due to error %s. To be completed with SRB_STATUS_BUSY. QueueNumber = %lu, SRB = 0x%p, Lun = %d, TimeOut = %d.\n", 
             (res == -1) ? "ENOSPC" : "UNKNOWN", QueueNumber, srbExt->Srb, SRB_LUN(Srb), Srb->TimeOutValue);

```
Any advice on how to proceed would be appreciated.

cc: @vrozenfe

**Comment by YanVugenfirer** [comment]
@benyamin-codez I will take a look. But I am travelling for the KVM Forum now and it might take me some time.

**Comment by JonKohler** [comment]
> @benyamin-codez I will take a look. But I am travelling for the KVM Forum now and it might take me some time. 

Glad you could make it. Travel logistics to Brno are brutal from where I live, I wanted to go this year but yikes, it's tough!

File: vioscsi/helper.c
Comment (benyamin-codez) [reply vioscsi/helper.c:112]: Per my comment below, we do actually set `-ENOSPC`, e.g.:

https://github.com/virtio-win/kvm-guest-drivers-windows/blob/22d0908a28f382d02d8d3a10a0ee0631af480fdf/VirtIO/VirtIORing.c#L263

**Review by vrozenfe — APPROVED** [review]

**Comment by benyamin-codez** [comment]
@vrozenfe
@YanVugenfirer 

Yan & Vadim, 

I've set this back to draft with a view to replacing it with a single commit PR.

I will rebase against master and add the changes I had reverted back in.

On the ENOSPC / SCSISTAT_QUEUE_FULL trace, I think it best to use:

```c
        RhelDbgPrint(TRACE_LEVEL_WARNING, 
                     " Could not put an SRB into a VQ due to error %s (%i). To be completed with SRB_STATUS_BUSY. QueueNumber = %lu, SRB = 0x%p, Lun = %d, TimeOut = %d.\n", 
                     (res == -ENOSPC) ? "ENOSPC" : "UNKNOWN", res, QueueNumber, srbExt->Srb, SRB_LUN(Srb), Srb->TimeOutValue);
```
Let me know if you want a PR to change the definition of `ENOSPC` in `\VirtIO\osdep.h` from 1 to 28 (or xyz) ...

cc: @JonKohler

**Comment by vrozenfe** [comment]
> 
> Let me know if you want a PR to change the definition of `ENOSPC` in `\VirtIO\osdep.h` from 1 to 28 (or xyz) ...
> 

As for me, I think it would be nice to keep the same definitions as in Linux.

Best,
Vadim.

**Comment by YanVugenfirer** [comment]
> > Let me know if you want a PR to change the definition of `ENOSPC` in `\VirtIO\osdep.h` from 1 to 28 (or xyz) ...
> 
> As for me, I think it would be nice to keep the same definitions as in Linux.
> 
> Best, Vadim.

@benyamin-codez I agree with Vadim. Better be consistence with Linux.

**Comment by benyamin-codez** [comment]
Single commit PR is #1162.

PR for `ENOSPC` change is #1163.

Closing this one...

**Pull request closed without merging**
